test/blackbox-tests/test-cases/pkg/ocamlformat/binary-avoid-from-project-deps.t

```diff
@@ -0,0 +1,164 @@
+Make sure that $ dune fmt does not take the OCamlFormat binary inside the "dune-project"
+dependencies.
+
+  $ . ../helpers.sh
+  $ mkrepo
+
+Make a fake ocamlformat.0.26.2:
+  $ mkdir ocamlformat
+  $ cd ocamlformat
+  $ cat > dune-project <<EOF
+  > (lang dune 3.13)
+  > (package (name ocamlformat))
+  > EOF
+  $ cat > ocamlformat.ml <<EOF
+  > let version = "0.26.2"
+  > ;;
+  > let () = print_endline ("formatted with version "^version)
+  > EOF
+  $ cat > dune <<EOF
+  > (executable
+  >  (public_name ocamlformat))
+  > EOF
+  $ cd ..
+  $ tar -czf ocamlformat-0.26.2.tar.gz ocamlformat
+
+Make a fake ocamlformat.0.26.3:
+  $ cd ocamlformat
+  $ cat > ocamlformat.ml <<EOF
+  > let version = "0.26.3"
+  > let () = print_endline ("formatted with version "^version)
+  > EOF
+  $ cd ..
+  $ tar -czf ocamlformat-0.26.3.tar.gz ocamlformat
+  $ rm -rf ocamlformat
+
+Make a package for the fake OCamlformat library:
+  $ mkpkg ocamlformat 0.26.2 <<EOF
+  > build: [
+  >   [
+  >     "dune"
+  >     "build"
+  >     "-p"
+  >     name
+  >     "@install"
+  >   ]
+  > ]
+  > url {
+  >  src: "file://$PWD/ocamlformat-0.26.2.tar.gz"
+  >  checksum: [
+  >   "md5=$(md5sum ocamlformat-0.26.2.tar.gz | cut -f1 -d' ')"
+  >  ]
+  > }
+  > EOF
+
+Make a package for the fake OCamlformat library:
+  $ mkpkg ocamlformat 0.26.3 <<EOF
+  > build: [
+  >   [
+  >     "dune"
+  >     "build"
+  >     "-p"
+  >     name
+  >     "@install"
+  >   ]
+  > ]
+  > url {
+  >  src: "file://$PWD/ocamlformat-0.26.3.tar.gz"
+  >  checksum: [
+  >   "md5=$(md5sum ocamlformat-0.26.3.tar.gz | cut -f1 -d' ')"
+  >  ]
+  > }
+  > EOF
+
+Make a project that uses the fake ocamlformat:
+  $ cat > dune-project <<EOF
+  > (lang dune 3.13)
+  > (package
+  >  (name foo)
+  >  (depends (ocamlformat (= 0.26.2))))
+  > EOF
+  $ cat > foo.ml <<EOF
+  > let () = print_endline "Hello, world"
+  > EOF
+  $ cat > dune <<EOF
+  > (executable
+  >  (public_name foo))
+  > EOF
+  $ cat > dune-workspace <<EOF
+  > (lang dune 3.13)
+  > (lock_dir
+  >  (path "dev-tools.locks/ocamlformat")
+  >  (repositories mock))
+  > (lock_dir
+  >  (repositories mock))
+  > (repository
+  >  (name mock)
+  >  (source "file://$(pwd)/mock-opam-repository"))
+  > EOF
+
+Add a fake binary in the PATH
+  $ mkdir .bin
+  $ cat > .bin/ocamlformat <<EOF
+  > #!/bin/sh
+  > echo "fake ocamlformat from PATH"
+  > EOF
+  $ chmod +x .bin/ocamlformat
+  $ PATH=$PWD/.bin:$PATH
+  $ which ocamlformat
+  $TESTCASE_ROOT/.bin/ocamlformat
+
+Lock and build the project to make OCamlFormat from project-dependencies available.
+  $ dune pkg lock
+  Solution for dune.lock:
+  - ocamlformat.0.26.2
+  $ dune fmt
```

Comment:
It might be worth running `dune fmt --preview` here instead. Dune will print out the diff of what change it would make without actually applying the change. This way you don't have to update "foo.ml" to contain its original contents every time you run `dune fmt` as the file won't be changed.